pkg/rbeconfigsgen/rbeconfigsgen.go

```diff
@@ -0,0 +1,847 @@
+	}
+	if _, err := io.Copy(o, resp.Body); err != nil {
+		return "", fmt.Errorf("error while downloading Bazelisk to %s: %w", localPath, err)
+	}
+
+	bazeliskContainerPath := path.Join(r.workdir, filename)
+	if err := r.copyTo(localPath, bazeliskContainerPath); err != nil {
+		return "", fmt.Errorf("failed to copy the downloaded Bazelisk binary into the container: %w", err)
+	}
+
+	if _, err := r.execCmd("chmod", "+x", bazeliskContainerPath); err != nil {
+		return "", fmt.Errorf("failed to mark the Bazelisk binary as executable inside the container: %w", err)
+	}
+	return bazeliskContainerPath, nil
+}
+
+func appendCppEnv(env []string, o *Options) ([]string, error) {
+	for k, v := range o.CppGenEnv {
+		env = append(env, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	if len(o.CppGenEnvJSON) == 0 {
+		return env, nil
+	}
+
+	blob, err := ioutil.ReadFile(o.CppGenEnvJSON)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read JSON file %q to read C++ config generation environment variables from: %w", o.CppGenEnvJSON, err)
+	}
+
+	e := map[string]string{}
+	if err := json.Unmarshal(blob, &e); err != nil {
+		return nil, fmt.Errorf("unable to parse file %q as a JSON string -> string dictionary: %w", o.CppGenEnvJSON, err)
+	}
+
+	for k, v := range e {
+		env = append(env, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	return env, nil
+}
+
+// genCppConfigs generates C++ configs inside the running toolchain container represented by the
+// given docker runner according to the given options. bazeliskPath is the path to the bazelisk
+// binary inside the running toolchain container.
+// The return value is the path to the C++ configs tarball copied out of the toolchain container.
+func genCppConfigs(r *dockerRunner, o *Options, bazeliskPath string) (string, error) {
+	if !o.GenCPPConfigs {
+		return "", nil
+	}
+
+	// Change the working directory to a dedicated empty directory for C++ configs for each
+	// command we run in this function.
+	cppProjDir := path.Join(r.workdir, "cpp_configs_project")
+	if _, err := r.execCmd("mkdir", cppProjDir); err != nil {
+		return "", fmt.Errorf("failed to create empty directory %q inside the toolchain container: %w", cppProjDir, err)
+	}
+	oldWorkDir := r.workdir
+	r.workdir = cppProjDir
+	defer func() {
+		r.workdir = oldWorkDir
+	}()
+
+	if _, err := r.execCmd("touch", "WORKSPACE", "BUILD.bazel"); err != nil {
+		return "", fmt.Errorf("failed to create empty build & workspace files in the container to initialize a blank Bazel repository: %w", err)
+	}
+
+	// Backup the current environment.
+	oldEnv := r.env
+	// Create a new environment for bazelisk commands used to specify the Bazel version to use to
+	// Bazelisk.
+	bazeliskEnv := []string{fmt.Sprintf("USE_BAZEL_VERSION=%s", o.BazelVersion)}
+	r.env = []string{fmt.Sprintf("USE_BAZEL_VERSION=%s", o.BazelVersion)}
+	// Always restore the old env before returning.
+	defer func() {
+		r.env = oldEnv
```

Comment:
wondering out loud: does it make sense to pass the env as a parameter to execCmd() instead of doing this dance?

Reply:
I considered it when writing this code and decided against it.

Rationale: There are 13 calls to execCmd and only two of them need to specify the env. Making the env a mandatory argument would add unnecessary placeholders for 11 calls. Can't use variadic args for env because I'm already using it for the command. So I preferred this dance because I only need it for two calls.